src/meta/meta_ckpt.c

```diff
@@ -503,42 +537,211 @@ __wt_meta_blk_mods_load(WT_SESSION_IMPL *session, const char *config, WT_CKPT *c
+          ((ckpt_saved != NULL && ckpt_saved->order == 0) &&
+            (ckpt_new != NULL && ckpt_new->order == 0))))
+        WT_ASSERT(session, false);
+}
+#endif
+
+/*
+ * __meta_ckptlist_allocate_new_ckpt --
+ *     Provided a checkpoint list, allocate a new checkpoint. Either use the last checkpoint in the
+ *     list or the file metadata to initialize this new checkpoint.
+ */
+static int
+__meta_ckptlist_allocate_new_ckpt(
+  WT_SESSION_IMPL *session, WT_CKPT **ckptbasep, size_t *allocated, const char *config)
+{
+    WT_CKPT *ckptbase, *ckpt;
+    WT_CONNECTION_IMPL *conn;
+    size_t slot;
+    uint64_t most_recent;
+
+    ckptbase = *ckptbasep;
+    conn = S2C(session);
+    slot = 0;
+
+    if (ckptbase != NULL)
+        WT_CKPT_FOREACH (ckptbase, ckpt)
+            slot++;
+
+    /* Either we have a configuration or an existing checkpoint to initialize with. */
+    WT_ASSERT(session, config != NULL || slot != 0);
+
+    /*
+     * If we are using an existing checkpoint, we must have the associated metadata. Otherwise we
+     * will have to go slow path and read the metadata.
+     */
+    if (config == NULL && ckptbase[slot - 1].block_metadata == NULL)
+        return (WT_NOTFOUND);
+
+    /*
+     * This isn't clean, but there's necessary cooperation between the schema layer (that maintains
+     * the list of checkpoints), the btree layer (that knows when the root page is written, creating
+     * a new checkpoint), and the block manager (which actually creates the checkpoint). All of that
+     * cooperation is handled in the array of checkpoint structures referenced from the WT_BTREE
+     * structure.
+     *
+     * Allocate a slot for a new value, plus a slot to mark the end.
+     */
+    WT_RET(__wt_realloc_def(session, allocated, slot + 2, &ckptbase));
+    *ckptbasep = ckptbase;
+
+    ckpt = &ckptbase[slot];
+    ckpt->order = (slot == 0) ? 1 : ckptbase[slot - 1].order + 1;
+    __wt_seconds(session, &ckpt->sec);
+    /*
+     * Update time value for most recent checkpoint, not letting it move backwards. It is possible
+     * to race here, so use atomic CAS. This code relies on the fact that anyone we race with will
+     * only increase (never decrease) the most recent checkpoint time value.
+     */
+    for (;;) {
+        WT_ORDERED_READ(most_recent, conn->ckpt_most_recent);
+        if (ckpt->sec <= most_recent ||
+          __wt_atomic_cas64(&conn->ckpt_most_recent, most_recent, ckpt->sec))
+            break;
+    }
+
+    /* Either load block mods from the config, or from the previous checkpoint. */
+    WT_RET(__wt_meta_blk_mods_load(
+      session, config, (slot == 0 ? NULL : &ckptbase[slot - 1]), ckpt, false));
+    WT_ASSERT(session, ckpt->block_metadata != NULL);
+
+    return (0);
+}
+
+/*
+ * __wt_meta_saved_ckptlist_get --
+ *     Append the ckptlist with a new checkpoint to be added.
+ */
+int
+__wt_meta_saved_ckptlist_get(WT_SESSION_IMPL *session, const char *fname, WT_CKPT **ckptbasep)
```

Comment:
I'm confused how we ended up with multiple flavors of "get the checkpoint list" function. I expected there would be a single "get the checkpoint list" function, that would look in the WT_BTREE handle first, and other wise generate it from the metadata. Can you talk a little bit about how you ended up with the separation you have?

Reply:
It is a portion of code I have never worked on before, so partially that is the reason. To be super safe, I started with adding a new function, that was a copy of the existing function `__wt_meta_ckptlist_get_from_config`, and instead of passing the config modified portions needed for it to work on the saved checkpoint list. I also started to call the original `__wt_meta_ckptlist_get` and compare the checkpoint lists and assert if the checkpoints differ. At the point where most functionalities worked, I put the common portions of the two functions in separate callable routines, so that there is no code duplication. While doing so, I did think about merging them into one function, which could either take `btree->ckpt` saved checkpoint list, or read metadata, similar to how some of the subroutines work now. I decided against a single such routine because I thought it will be safer to keep `__wt_meta_saved_ckptlist_get` as a separate function, callable only from `__checkpoint_lock_dirty_tree`. `__wt_meta_ckptlist_get` also gets called as part of verify, `__wt_meta_ckptlist_get_from_config` also gets called for import and repair. Not knowing well how those functionalities will be impacted, I decided to keep the functions separate.

In short, a lot of code structure ended up like this because I was trying to be cautious, and trying to keep those existing code workflows that won't be needing a saved checkpoint list unimpacted.

Reply:
Intuitively, it seems to me that no code other than the underlying metadata code should should know where the checkpoint information came from, since the two sources should have identical results. Unless you disagree, I'd seriously consider re-factoring so there's a single "get the checkpoint info" call.

Reply:
I agree that refactoring is desirable. Feel free to make it a separate ticket with clear instructions about what exists and what should exist.

Reply:
I have created WT-7524 to refactor. Marking the comment resolved for now. I have put the refactor tickets as part of the project and should be able to pick them up very soon.